tokio/src/future/poll_fn.rs

```diff
@@ -4,22 +4,25 @@
 
 use std::fmt;
 use std::future::Future;
+use std::marker::PhantomPinned;
 use std::pin::Pin;
 use std::task::{Context, Poll};
 
 /// Future for the [`poll_fn`] function.
 pub struct PollFn<F> {
     f: F,
+    _pinned: PhantomPinned,
```

Comment:
Why make it *always* `!Unpin`?  That looks like an overreaction, I'd recommend to to the same as https://github.com/rust-lang/rust/pull/102737.

Reply:
I understand that it isn't necessary, but I don't see why not.

Reply:
I find it makes the code harder to read, like code that uses SeqCst when release/acquire would have been sufficient -- I'll go try to hunt down the reason for *why* the code need this strong guarantee, when there actually is no reason for it.

Reply:
(Btw I am not sure if I understood your comment correctly. Usually I would have parsed "I don't see why not" as "I don't understand why not", but that contradicts the first part of your sentence, so I was left confused. Maybe the "not" did not refer to "not necessary" -- I am unsure.)

Reply:
I'm pretty sure that you understood me correctly.

Reply:
This is the reason that I did it this way originally: Implementing your suggestion is done by doing nothing, and having `Unpin` be implemented implicitly. I preferred a solution that actually explicitly did something. However, I have now applied your suggestion and added a test for it being `!Unpin` when it should.